### PR TITLE
Dangling example reference autofill facility

### LINKED ISSUE #1845: Support filling out missing references with plausible examples
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

A very common problem in providers like AWS are incomplete HCL examples that assume some missing context and as a result do not compile or preview, specifically often the examples assume a resource that is not listed in the example. It turns out that in TF/HCL representation the types of the missing resources are fairly obvious. It would help out out a lot if we could supply a folder with "canonical" definitions of resources indexed by type, and teach the converter to automatically augment the examples.

One prominent AWS concrete example I'm looking at:

```terraform
 resource "aws_route53_record" "example" {
      for_each = {
        for dvo in aws_acm_certificate.example.domain_validation_options : dvo.domain_name => {
          name   = dvo.resource_record_name
          record = dvo.resource_record_value
          type   = dvo.resource_record_type
        }
      }

      allow_overwrite = true
      name            = each.value.name
      records         = [each.value.record]
      ttl             = 60
      type            = each.value.type
      zone_id         = aws_route53_zone.example.zone_id
}

### Inject this example after we find a missing reference to aws_acm_certificate
resource "aws_acm_certificate" "example" {
  domain_name       = "example.com"
  validation_method = "DNS"
}

### Inject this example after we find a missing reference to aws_route53_zone
resource "aws_route53_zone" "example" {
  name = "example.com"
}
```

I almost got it to preview.  There is some other bug where I had to inject `<any>` to make the type checker happy `       type: (<any>aws.route53.RecordType)[range.value.type],`.

```
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const exampleCertificate = new aws.acm.Certificate("example", {
    domainName: "example.com",
    validationMethod: "DNS",
});
const exampleZone = new aws.route53.Zone("example", {name: "example.com"});
const example: aws.route53.Record[] = [];
exampleCertificate.domainValidationOptions.apply(domainValidationOptions => {
    for (const range of Object.entries(domainValidationOptions.reduce((__obj, dvo) => ({ ...__obj, [dvo.domainName]: {
        name: dvo.resourceRecordName,
        record: dvo.resourceRecordValue,
        type: dvo.resourceRecordType,
    } }))).map(([k, v]) => ({key: k, value: v}))) {
        example.push(new aws.route53.Record(`example-${range.key}`, {
            allowOverwrite: true,
            name: range.value.name,
            records: [range.value.record],
            ttl: 60,
            type: (<any>aws.route53.RecordType)[range.value.type],
            zoneId: exampleZone.zoneId,
        }));
    }
});
```

### PR DESCRIPTION
**Pull request by t0yv0**
Fixes #1845 

Introduce a facility that automatically corrects HCL snippets by injecting resource definitions for the referenced but omitted resources. These auto definitions are taken from a provider-author-supplied dictionary of "vanilla" definitions that is indexed by type of resource. Facilities are provided to build one from a folder with `my_resource.tf` files.

For example in AWS it make this example compile and fixes 18 other examples:

```typescript

  * import * as aws from "@pulumi/aws";
  *
+ * const exampleCertificate = new aws.acm.Certificate("example", {
+ *     domainName: "example.com",
+ *     validationMethod: "DNS",
+ * });
+ * const exampleZone = new aws.route53.Zone("example", {name: "example.com"});
  * const example: aws.route53.Record[] = [];
- * for (const range of Object.entries(.reduce((__obj, dvo) => ({ ...__obj, [dvo.domainName]: {
- *     name: dvo.resourceRecordName,
- *     record: dvo.resourceRecordValue,
- *     type: dvo.resourceRecordType,
- * } }))).map(([k, v]) => ({key: k, value: v}))) {
- *     example.push(new aws.route53.Record(`example-${range.key}`, {
- *         allowOverwrite: true,
- *         name: range.value.name,
- *         records: [range.value.record],
- *         ttl: 60,
- *         type: aws.route53.RecordType[range.value.type],
- *         zoneId: exampleAwsRoute53Zone.zoneId,
- *     }));
- * }
+ * exampleCertificate.domainValidationOptions.apply(domainValidationOptions => {
+ *     for (const range of Object.entries(domainValidationOptions.reduce((__obj, dvo) => ({ ...__obj, [dvo.domainName]: {
+ *         name: dvo.resourceRecordName,
+ *         record: dvo.resourceRecordValue,
+ *         type: dvo.resourceRecordType,
+ *     } }))).map(([k, v]) => ({key: k, value: v}))) {
+ *         example.push(new aws.route53.Record(`example-${range.key}`, {
+ *             allowOverwrite: true,
+ *             name: range.value.name,
+ *             records: [range.value.record],
+ *             ttl: 60,
+ *             type: aws.route53.RecordType[range.value.type],
+ *             zoneId: exampleZone.zoneId,
+ *         }));
+ *     }
+ * });

```

All I had to do is to populate a few files in AWS:

```
resource "aws_acm_certificate" "example" {
  domain_name       = "example.com"
  validation_method = "DNS"
}
```

and:

```
resource "aws_route53_zone" "example" {
  name = "example.com"
}
```

### CODE REVIEW

**Comment by t0yv0** [comment]
Hmm. This is pulling in pulumi/terraform fork as a dependency which will transitively impact all providers - to access the TF parser goodness. I recall we were trying to avoid this specifically linking to the TF converter via a process boundary. Perhaps this can be done as a converter feature after all.

**Comment by t0yv0** [comment]
Isolated HCL parser dependency into a tool so that it's not propagated to the repositories depending on the bridge.

**Review by iwahbe — APPROVED** [review]
This is an interesting solution, and will help us achieve the goal of *copy & paste* examples.

Before merging, it would be great to have some documentation. I think we need:
- A `README.md` in `tools/pulumi-hcl-lint` briefly explaining what the tool does, and how it is intended to be consumed. 
- A page in `docs`: `docs/dangling-reference-autofill.md` that explains who this feature is for (all provider authors?, in beta and just for Pulumi?, ect.) and how it should be consumed.

File: tools/go.mod
Comment (t0yv0) [reply tools/go.mod:1]: I don't know that that's necessary - I think we might be erring on proliferating too many go.mod files here already compared to what other companies do.  I was going with https://github.com/golang-standards/project-layout/tree/master/tools

Looks like the examples they list use one global go.mod at the root. 

We don't do it here because of the ask not to proliferate some of the dependencies into the core libraries that make all bridged providers depend on them as well. Hence a dedicated go.mod for tools.

Do we care to isolate dependencies from one tool to another? I can't think of why. Perhaps we can cross that bridge when we get to it.

Comment (iwahbe) [reply tools/go.mod:1]: I don't feel strongly here, so don't block on this. It feels pretty odd to me that `tools` has it's own `go.mod` (vs each tool having one), but that might go against standard go practice.

File: tools/pulumi-hcl-lint/lint_dangling_refs.go
Comment (VenelinMartinov) [inline tools/pulumi-hcl-lint/lint_dangling_refs.go:35]: This is for loop variables, right? They are assumed to be defined?

**Review by VenelinMartinov — APPROVED** [review]
lgtm, thanks for adding tests on both sides.

File: tools/pulumi-hcl-lint/lint_dangling_refs.go
Comment (t0yv0) [reply tools/pulumi-hcl-lint/lint_dangling_refs.go:35]: This came up as a false positive dangling ref detected in the one test example I was using.

**Pull request merged**
